storybookjs/storybook
# Core: Add 'staticDirs' config option

Issue: #13250

## ☑️ What I did
- [x] Added `staticDirs` property to  `StorybookConfig` type
- [x] Added missing type to `useStatics(...)` second argument `options` as `Options` from `@storybook/core-common`
- [x] `useStatics` reads `staticDirs` property from Storybook's configuration and, if it's defined, it adds a route for each directory path passed to staticDirs:
   - The property supports a `string` or an object `{from: string, to: string}`
   - if it's a string, the route of the server for those statics should be `/` and serve the files from `${staticDir}`
   - if it's an object, the route of the server for those statics should be `/${to}/` and serve the files from `${from}`
- [x] Deprecate `--staticDir` | `-s` option from Storybook CLI
   - [x] Mark properties defined in types as `/* deprecated */` (`prod` and `dev`)
   - [x] Add `--static-dir` flag to the collection of flags which should trigger an warning when running the CLI

## 🔬 How to test
### Manual testing
1. Choose one storybook framework example you want to run the manual test with _i.e. `examples/react-ts`_
2. Create a new folder with the name you want under the storybook example you chose to run the manual test. Create a file inside the folder you just created with the filename you want and the content you want. i.e. `examples/react-ts/public/public.txt`
3. Change the `main.ts` of the example you selected to run the test adding the `staticDirs` property and set the value of the folder you created in the previous step: i.e. 
```js
...
staticDirs: ['./public']
...
```
4. Run Storybook for the example you chose (make sure you do it in its corresponding folder). i.e.
```bash
> pwd
~/storybook/examples/react-ts

> yarn storybook
```
5. Once Storybook is running in your `localhost`, try to access your static file. _i.e. `http://localhost:<YOUR_PORT>/public.txt`_
![image](https://user-images.githubusercontent.com/2574275/131889401-f801269b-8e23-45b9-8315-4fcc55a69451.png)

6. Repeat all the process using the second approach of `{from, to}` and check your static files are served under its mapped path. i.e. `{ from: './public', to: './assets' }` -> `http://localhost:<PORT>/assets/test.txt`
![image](https://user-images.githubusercontent.com/2574275/131889286-3f979ac1-3726-41ff-b88d-5037fdc62286.png)

### [🚧  WIP 🚧 ] Non-manual Tests
- [ ] Add test coverage for this new Storybook configuration option
    - [x] Evaluate [edge cases covered by previous CLI's flag: `--staticDir` tests](https://github.com/storybookjs/storybook/blob/next/lib/core-server/src/utils/__tests__/server-statics.test.ts)
    - [x] Add `staticDirs` option to `examples/official-storybook/main.ts`

The modified method has no test coverage but I would like to use this PR to add some coverage into it. I would like to test the express app using some library like [`supertest`](url). Is there any chosen library to test this kind of thins in storybook repo I should be aware of?

## Documentation
- [x] Add documentation about this new Storybook configuration option
- [x] Update documentation with deprecated CLI flag
